### PR TITLE
TF-2387 Add starting page for mobile

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2387 

## DoD

- [x] Add Starting page
- [x] Handle `Use composer server` button
- [x] Navigate starting page when invoke logout action
- [x] Handle `Create Twake ID` button
- [x] Handle `Sign in` button
- [x] Handle `Logout` with twake workplace

## Resolved

- Sign In with SaaS:

[demo-sign-in-saas.webm](https://github.com/user-attachments/assets/43b25924-4955-4f1a-9cb0-3cfdf47fc2a3)

- Sign Up with Twake Workplace:

[demo-signup-saas.webm](https://github.com/user-attachments/assets/1d11e978-02c1-45ab-a16b-9a628344b613)

- Sign In via Sign Up flow:

[demo-sign-in-via-sign-up.webm](https://github.com/user-attachments/assets/0f3562db-7890-405f-b31d-aafb0252d55d)

- Sign In via company server:

 
[sign-in-normal.webm](https://github.com/user-attachments/assets/8e255896-d41d-4c1a-ade2-281ecaf6773a)

### CODE REVIEW

**Comment by hoangdat** [comment]
- new design: https://www.figma.com/design/uLw3iWnOEGaRlJnfPz4Pv0/Twake-%5BMatrix%5D?node-id=28813-8600&t=6WN1GjmkzLRKDqPY-4

**Comment by dab246** [comment]
> * new design: https://www.figma.com/design/uLw3iWnOEGaRlJnfPz4Pv0/Twake-%5BMatrix%5D?node-id=28813-8600&t=6WN1GjmkzLRKDqPY-4

I can't access it. Please approve my email. Thanks.

**Comment by dab246** [comment]
## Blocker

### Description

- When I register an account on TW by authenticating to the url `https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=twake.tmail://redirect&app=tmail`, after successful registration the system returns me a Uri of `twake.tmail://redirect`.

### Propose

- We should return the token when sign up is successful.


### Video demo

[demo.webm](https://github.com/user-attachments/assets/dd24b115-2bea-450b-93b3-47c96d7c6f8f)

**Comment by rezk2ll** [comment]
> ## Blocker
> ### Description
> * When I register an account on TW by authenticating to the url `https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=twake.tmail://redirect&app=tmail`, after successful registration the system returns me a Uri of `twake.tmail://redirect`.
> 
> ### Propose
> * We should return the token when sign up is successful.
> 
> ### Video demo
>  demo.mov 
> @hoangdat

what type of token is required? ( loginToken, access_token...) ?

**Comment by hoangdat** [comment]
accessToken and refreshToken

**Comment by hoangdat** [comment]
hi @rezk2ll , any update on this topic?

**Comment by rezk2ll** [comment]
> * https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=twake.tmail://redirect&app=tmail

it's live on staging now @hoangdat @dab246 

### sign-up from tmail:

```
https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=twake.tmail://redirect&app=tmail
```

### login from tmail:

```
https://sign-up.stg.lin-saas.com/?post_login_redirect_url=twake.tmail://redirect&app=tmail
```

### result:

after sign-up / login success redirect to

```
twake.tmail://redirect?access_token=ey....
```

**Comment by dab246** [comment]
> > * https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=twake.tmail://redirect&app=tmail
> 
> it's live on staging now @hoangdat @dab246
> 
> ### sign-up from tmail:
> ```
> https://sign-up.stg.lin-saas.com/?post_registered_redirect_url=teammail.mobile://redirect&app=tmail
> ```
> 
> ### login from tmail:
> ```
> https://sign-up.stg.lin-saas.com/?post_login_redirect_url=teammail.mobile://redirect&app=tmail
> ```
> 
> ### result:
> after sign-up / login success redirect to
> 
> ```
> twake.tmail://redirect?access_token=ey....&refresh_token=something&id_token=toke&expires_in=3600
> ```
> 
> ### demo
>  tmail-redirect-oidc.mp4

@rezk2ll  Thanks for the implementation. But since we use another authenticate library to authenticate the registrations, we need a different `url scheme` than the current `teammail.mobile`. You can name it `twake.tmail` for example.

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      super.handleFailureViewState(failure);
+    }
+  }
+
+  @override
+  void handleReloaded(Session session) {
+    TipDialogHelper.dismiss();
+
+    popAndPush(
+      RouteUtils.generateNavigationRoute(AppRoutes.dashboard),
```
Comment (tddang-linagora) [inline lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart:131]: Why do we navigate to another screen on reload?

File: lib/features/starting_page/domain/usecase/sign_in_twake_workplace_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  }) async* {
+    try {
+      yield Right<Failure, Success>(SignInTwakeWorkplaceLoading());
+
+      final tokenOIDC = await _saasRepository.signInTwakeWorkplace(oidcConfiguration);
+
+      await Future.wait([
+        _credentialRepository.saveBaseUrl(baseUri),
+        _authenticationOIDCRepository.persistTokenOIDC(tokenOIDC),
+        _authenticationOIDCRepository.persistAuthorityOidc(oidcConfiguration.authority),
```
Comment (tddang-linagora) [inline lib/features/starting_page/domain/usecase/sign_in_twake_workplace_interactor.dart:39]: If one of these throws, `SignInTwakeWorkplaceSuccess` would not be returned. Is this expected behavior?

File: lib/features/starting_page/domain/usecase/sign_up_twake_workplace_interactor.dart [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  }) async* {
+    try {
+      yield Right<Failure, Success>(SignUpTwakeWorkplaceLoading());
+
+      final tokenOIDC = await _saasRepository.signUpTwakeWorkplace(oidcConfiguration);
+
+      await Future.wait([
+        _credentialRepository.saveBaseUrl(baseUri),
+        _authenticationOIDCRepository.persistTokenOIDC(tokenOIDC),
+        _authenticationOIDCRepository.persistAuthorityOidc(oidcConfiguration.authority),
```
Comment (tddang-linagora) [inline lib/features/starting_page/domain/usecase/sign_up_twake_workplace_interactor.dart:39]: idem

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+class TwakeWelcomeView extends GetWidget<TwakeWelcomeController> {
+
+  const TwakeWelcomeView({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    SystemChrome.setPreferredOrientations([
+      DeviceOrientation.portraitUp,
+      DeviceOrientation.portraitDown
+    ]);
```
Comment (tddang-linagora) [inline lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart:21]: If our app supports landscape mobile mode, make sure to reset this config when this screen closes.

Comment (dab246) [reply lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart:21]: It must be so.

File: lib/features/starting_page/domain/usecase/sign_in_twake_workplace_interactor.dart (hunk 2 above)
Comment (dab246) [reply lib/features/starting_page/domain/usecase/sign_in_twake_workplace_interactor.dart:39]: Of course, what do you want here?

File: lib/features/starting_page/domain/usecase/sign_up_twake_workplace_interactor.dart (hunk 3 above)
Comment (dab246) [reply lib/features/starting_page/domain/usecase/sign_up_twake_workplace_interactor.dart:39]: Of course, what do you want here?

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart:131]: This is the function called when session is successfully retrieved.

File: lib/features/login/presentation/login_controller.dart [hunk 5]
```diff
@@ -254,20 +255,30 @@ class LoginController extends ReloadableController {
     ));
   }
 
-  void handleBackButtonAction() {
+  void handleBackButtonAction(BuildContext context) {
+    KeyboardUtils.hideKeyboard(context);
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/login/presentation/login_controller.dart:259]: Either way is fine, We have context, we should hide keyboard in that context.

File: lib/features/login/presentation/login_controller.dart [hunk 6]
```diff
@@ -277,7 +288,8 @@ class LoginController extends ReloadableController {
     userNameFocusNode.requestFocus();
   }
 
-  void handleLoginPressed() {
+  void handleLoginPressed(BuildContext context) {
+    KeyboardUtils.hideKeyboard(context);
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/login/presentation/login_controller.dart:292]: Either way is fine, We have context, we should hide keyboard in that context.

**Review by tddang-linagora — DISMISSED** [review]

**Comment by dab246** [comment]
Done. @hoangdat Please review it.

**Review by tddang-linagora — DISMISSED** [review]

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart [hunk 7]
```diff
@@ -25,7 +31,29 @@ extension OidcConfigurationExtensions on OIDCConfiguration {
...
     } else {
-      return redirectOidcMobile;
-    }
+      return AppConfig.isSaasPlatForm
+        ? OIDCConstant.twakeWorkplaceRedirectUrl
+        : redirectOidcMobile;
+  }
   }
+
+  String get singInUrl => ServicePath(authority)
```
Comment (hoangdat) [inline lib/features/login/domain/extensions/oidc_configuration_extensions.dart:40]:
```suggestion
  String get singInTWPUrl => ServicePath(authority)
```

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart [hunk 8]
```diff
@@ -25,7 +31,29 @@ extension OidcConfigurationExtensions on OIDCConfiguration {
...
+    .withQueryParameters([
+      StringQueryParameter(
+        OIDCConstant.postLoginRedirectUrlPathParams,
+        OIDCConstant.twakeWorkplaceRedirectUrl,
+      ),
+      StringQueryParameter('app', OIDCConstant.appParameter),
+    ])
+    .generateEndpointPath();
+
+  String get signUpUrl => ServicePath(authority)
```
Comment (hoangdat) [inline lib/features/login/domain/extensions/oidc_configuration_extensions.dart:50]:
```suggestion
  String get signUpTWPUrl => ServicePath(authority)
```

File: env.file [hunk 9]
```diff
@@ -6,4 +6,6 @@ APP_GRID_AVAILABLE=supported
 FCM_AVAILABLE=supported
 IOS_FCM=supported
 FORWARD_WARNING_MESSAGE=
-PLATFORM=other
\ No newline at end of file
+PLATFORM=other
+REGISTRATION_URL=http://localhost/
```
Comment (hoangdat) [inline env.file:10]:
```suggestion
SAAS_REGISTRATION_URL=http://localhost/
```

File: env.file [hunk 10]
```diff
@@ -6,4 +6,6 @@ APP_GRID_AVAILABLE=supported
...
+SAAS_SERVER_URL=http://localhost/
```
Comment (hoangdat) [inline env.file:11]:
```suggestion
SAAS_JMAP_SERVER_URL=http://localhost/
```

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+          focusColor: Colors.transparent,
+          hoverColor: Colors.transparent,
+          highlightColor: Colors.transparent,
+          overlayColor: WidgetStateProperty.all(Colors.transparent),
+          signInTitle: AppConfig.isSaasPlatForm
+            ? AppLocalizations.of(context).signIn.capitalizeFirst ?? ''
+            : null,
+          onSignInOnTap: () => controller.onClickSignIn(context),
+          onCreateTwakeIdOnTap: () => controller.onSignUpTwakeWorkplace(context),
+          createTwakeIdTitle: AppConfig.isSaasPlatForm
```
Comment (hoangdat) [inline lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart:42]: no need, please show signIn/signUp button by default

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart (hunk 7 above)
Comment (dab246) [reply lib/features/login/domain/extensions/oidc_configuration_extensions.dart]: done

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart (hunk 8 above)
Comment (dab246) [reply lib/features/login/domain/extensions/oidc_configuration_extensions.dart]: done

File: env.file (hunk 9 above)
Comment (dab246) [reply env.file]: done

File: env.file (hunk 10 above)
Comment (dab246) [reply env.file]: done

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart (hunk 11 above)
Comment (dab246) [reply lib/features/starting_page/presentation/twake_welcome/twake_welcome_view.dart]: done

**Review by tddang-linagora — DISMISSED** [review]

File: lib/features/base/base_controller.dart [hunk 12]
```diff
@@ -407,14 +407,32 @@ abstract class BaseController extends GetxController
...
+    popAndPush(AppRoutes.twakeWelcome);
+  }
+
+  void navigateToLoginPage() {
+    if (Get.currentRoute == AppRoutes.login) {
+      return;
     }
+    pushAndPopAll(
+      AppRoutes.login,
+      arguments: LoginArguments(LoginFormType.none));
```
Comment (hoangdat) [inline lib/features/base/base_controller.dart:435]: why not ?

```suggestion
      arguments: PlatformInfo.isMobile ? LoginFormType.dnsLookupForm : LoginFormType.none
```

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart [hunk 13]
```diff
@@ -1,19 +1,23 @@
...
-        return AppConfig.domainRedirectUrl + loginRedirectOidcWeb;
-      } else {
-        return '${AppConfig.domainRedirectUrl}/$loginRedirectOidcWeb';
-      }
+      return AppConfig.domainRedirectUrl.endsWith('/')
+        ? AppConfig.domainRedirectUrl + loginRedirectOidcWeb
+        : '${AppConfig.domainRedirectUrl}/$loginRedirectOidcWeb';
     } else {
-      return redirectOidcMobile;
+      return authority == AppConfig.saasRegistrationUrl
```
Comment (hoangdat) [inline lib/features/login/domain/extensions/oidc_configuration_extensions.dart:18]: Should use function
```
bool isSaasAuthority(authority) 
```

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart [hunk 14]
```diff
@@ -25,7 +29,29 @@ extension OidcConfigurationExtensions on OIDCConfiguration {
         return '${AppConfig.domainRedirectUrl}/$logoutRedirectOidcWeb';
       }
     } else {
-      return redirectOidcMobile;
-    }
+      return authority == AppConfig.saasRegistrationUrl
```
Comment (hoangdat) [inline lib/features/login/domain/extensions/oidc_configuration_extensions.dart:32]: idem

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart [hunk 15]
```diff
@@ -25,7 +29,29 @@ extension OidcConfigurationExtensions on OIDCConfiguration {
...
+
+  String get signUpTWPUrl => ServicePath(authority)
+    .withQueryParameters([
+      StringQueryParameter(
+        OIDCConstant.postRegisteredRedirectUrlPathParams,
+        OIDCConstant.twakeWorkplaceRedirectUrl,
+      ),
+      StringQueryParameter('app', OIDCConstant.appParameter),
+    ])
+    .generateEndpointPath();
```
Comment (hoangdat) [inline lib/features/login/domain/extensions/oidc_configuration_extensions.dart:56]: should have a test to document the form of this URL

File: model/lib/oidc/token_oidc.dart [hunk 16]
```diff
@@ -27,6 +28,38 @@ class TokenOIDC with EquatableMixin {
...
+
+    final idToken = queryParams['id_token'];
+    if (idToken == null || idToken.isEmpty) {
+      throw TokenIdIsNullException();
+    }
+
+    final expiresIn = queryParams['expires_in'];
+    if (expiresIn == null || expiresIn.isEmpty) {
+      throw ExpiresTimeIsNullException();
+    }
```
Comment (hoangdat) [inline model/lib/oidc/token_oidc.dart:53]: No need to verify this, these are nullable fields which depend on scope of `token` request.

File: model/test/oidc/token_oidc_test.dart [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+      const uriString = 'https://example.com/callback?access_token=valid_access&refresh_token=valid_refresh&expires_in=3600';
+
+      expect(() => TokenOIDC.fromUri(uriString), throwsA(isA<TokenIdIsNullException>()));
+    });
+
+    test('throws ExpiresTimeIsNullException if expires_in is missing', () {
+      const uriString = 'https://example.com/callback?access_token=valid_access&refresh_token=valid_refresh&id_token=valid_id';
+
+      expect(() => TokenOIDC.fromUri(uriString), throwsA(isA<ExpiresTimeIsNullException>()));
+    });
```
Comment (hoangdat) [inline model/test/oidc/token_oidc_test.dart:39]: remove these tests

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  void onClickSignIn(BuildContext context) {
+    TipDialogHelper.loading(AppLocalizations.of(context).loadingPleaseWait);
+
+    final baseUri = Uri.tryParse(AppConfig.saasJmapServerUrl);
+
+    if (baseUri == null) {
+      consumeState(Stream.value(Left(SignInTwakeWorkplaceFailure(SaasServerUriIsNull()))));
+      return;
```
Comment (hoangdat) [inline lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart:53]: do we need to close Tip dialog here?

File: lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  void onSignUpTwakeWorkplace(BuildContext context) {
+    TipDialogHelper.loading(AppLocalizations.of(context).loadingPleaseWait);
+
+    final baseUri = Uri.tryParse(AppConfig.saasJmapServerUrl);
+
+    if (baseUri == null) {
+      consumeState(Stream.value(Left(SignUpTwakeWorkplaceFailure(SaasServerUriIsNull()))));
+      return;
```
Comment (hoangdat) [inline lib/features/starting_page/presentation/twake_welcome/twake_welcome_controller.dart:73]: idem

**Review by tddang-linagora — DISMISSED** [review]

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart (hunk 13 above)
Comment (dab246) [reply lib/features/login/domain/extensions/oidc_configuration_extensions.dart:18]: done

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart (hunk 14 above)
Comment (dab246) [reply lib/features/login/domain/extensions/oidc_configuration_extensions.dart:32]: done

File: model/lib/oidc/token_oidc.dart (hunk 16 above)
Comment (dab246) [reply model/lib/oidc/token_oidc.dart:53]: done

File: lib/features/login/domain/extensions/oidc_configuration_extensions.dart (hunk 15 above)
Comment (dab246) [reply lib/features/login/domain/extensions/oidc_configuration_extensions.dart:59]: done

**Review by hoangdat — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**